### PR TITLE
Server config refactor

### PR DESCRIPTION
**Pull request by apfong**
#### API Changes:

Changed the ServerConfig class to construct both maps of path to
config property value and prefix to handler names. A ServerConfig
object is initialized using Build() so that we can catch any errors
that might occur from reading a malformed config file.

Also introduced are two new getters, `allPaths()` and `getChildBlock()`
which return the mapping of prefixes to handler name and inner
NginxConfig block of the specified route config respectively.

#### Other Changes:

Refactored `fillOutMaps()` so that we no longer make use of exceptions
when checking if a property exists in a map. Instead we use
`unordered_map::find()`. This resolves Issue #48.

### CODE REVIEW

**Comment by KyCodeHuynh** [comment]
From offline discussion: Travis builds will continue to fail until RequestHandler tests are updated, and all our work is merged together in the `lightning-refactor` branch.

File: server_config_test.cc [hunk 1]
```diff
@@ -20,130 +21,243 @@ class ServerConfigTest : public ::testing::Test {
...
 
+TEST_F(ServerConfigTest, BuildOK) {
+    std::unique_ptr<ServerConfig> server_config
+      = ParseString("port 8080; \
+                     path /echo EchoHandler {}");
+    bool ec = server_config->Build();
+
+    EXPECT_EQ(true, ec)
+      << "Build() should return true if a valid config is supplied to a \
+          Server Config is supplied";
```
Comment (KyCodeHuynh) [inline server_config_test.cc]: Minor copy-paste typo: "Build() should return true if a valid config is supplied to a Server Config"

File: server_config_test.cc [hunk 2]
```diff
@@ -20,130 +21,243 @@ class ServerConfigTest : public ::testing::Test {
...
+    std::unique_ptr<ServerConfig> server_config
+      = ParseString("port 8080; \
+                     path /echo EchoHandler {} \
+                     default NotFoundHandler {}");
+    bool ec = server_config->Build();
+    std::unordered_map<std::string, std::string> paths = server_config->allPaths();
+
+    EXPECT_EQ(true, ec)
+      << "Build() should return true if a valid config is supplied";
+    EXPECT_EQ(false, paths.empty())
```
Comment (KyCodeHuynh) [inline server_config_test.cc]: Can we also check the size of the map?

File: server_config_test.cc [hunk 3]
```diff
@@ -20,130 +21,243 @@ class ServerConfigTest : public ::testing::Test {
...
+    EXPECT_EQ(0, childBlock->statements_.size())
+      << "The child block of '/' prefix should have no statements";
+}
+
 TEST_F(ServerConfigTest, NoPortConfig) {
     std::unique_ptr<ServerConfig> server_config = ParseString("server { ACM Hack; }");
-    std::vector<std::string> query = {"server", "listen"};
+    bool buildCode = server_config->Build();
+
+    std::vector<std::string> query = {"server", "port"};
```
Comment (KyCodeHuynh) [inline server_config_test.cc]: Could we place the test code + accompanying assertions together? That way, if we have more assertions later on, we don't have to scan up-and-down to pair up test-case with assertion.

For example: 

```c++
const auto actual_result1 = testCase1();
EXPECT_EQ(expected, actual_result);

const auto actual_result2 = testCase2();
EXPECT_EQ(expected, actual_result);
```

As compared to: 

```c++
const auto actual_result1 = testCase1();
const auto actual_result2 = testCase2();

EXPECT_EQ(expected, actual_result1);
EXPECT_EQ(expected, actual_result2);
```

Comment (KyCodeHuynh) [inline server_config_test.cc]: The same applies to the remaining test cases below.

File: server_config_test.cc [hunk 4]
```diff
@@ -20,130 +21,243 @@ class ServerConfigTest : public ::testing::Test {
...
 }
 
 TEST_F(ServerConfigTest, DoublyNestedPort) {
     std::unique_ptr<ServerConfig> server_config
-      = ParseString("outer { server { listen 1234; } }");
-    std::vector<std::string> query = {"outer", "server", "listen"};
+      = ParseString("outer { server { port 1234; } }");
+    bool buildCode = server_config->Build();
+
+    std::vector<std::string> query = {"outer", "server", "port"};
```
Comment (KyCodeHuynh) [inline server_config_test.cc]: Could we add a test case for where the query-string is missing the `outer` (and should fail)? That would let us guard against the regression where we inadvertently introduce an assumed outer scope, for example.

File: server_config.cc [hunk 5]
```diff
@@ -41,83 +34,156 @@ std::string buildWordyParam(std::shared_ptr<NginxConfigStatement> statement,
...
+  return fillOutMaps(config_, basePath);
+}
+
 // In NgnixConfig, Statements are root level 'blocks' (e.g. server {})
-// fillOutMap will recursively construct an unordered_map between a vector of strings
+// fillOutMaps will recursively construct an unordered_map between a vector of strings
 // representing the 'path' to a config parameter and the value of that parameter
 //   e.g. ["server", "listen"] should be mapped to the port number
-void ServerConfig::fillOutMap(NginxConfig config, std::vector<std::string> basePath) {
+bool ServerConfig::fillOutMaps(NginxConfig config, std::vector<std::string> basePath) {
```
Comment (KyCodeHuynh) [inline server_config.cc]: Very minor: can we standardize on `snake_case` for variables and parameters (as we currently have elsewhere), `lowerCamelCase` for methods, and `UpperCamelCase` for classes/namespaces/etc.?

File: server_config.cc [hunk 6]
```diff
@@ -41,83 +34,156 @@ std::string buildWordyParam(std::shared_ptr<NginxConfigStatement> statement,
...
     }
-    else if (config.statements_[i]->child_block_ != nullptr) {
+    else if (curStatement->child_block_ != nullptr) {
       // Dealing with an NginxConfig Block
-      std::vector<std::string> basePathExtended = basePath;
-      basePathExtended.push_back(buildWordyParam(config.statements_[i], 0));
-      fillOutMap(*(config.statements_[i]->child_block_), basePathExtended);
+
+      // If we're looking at a Route configuration block, stop recursing and
+      // store its child block in the prefix_to_handler_name_ map.
```
Comment (KyCodeHuynh) [inline server_config.cc]: Do we mean: "store [the handler name] in the prefix_to_handler_name_ map"? There is no child-block here?

File: server_config.cc [hunk 7]
```diff
@@ -41,83 +34,156 @@ std::string buildWordyParam(std::shared_ptr<NginxConfigStatement> statement,
...
-  // unordered_map::at(key) throws if key not found.
-  // unordered_map::operator[key] will silently create that entry
-  // See: http://www.cplusplus.com/reference/unordered_map/unordered_map/operator[]/
-  // and: http://www.cplusplus.com/reference/stdexcept/out_of_range/
-  try {
-    // TODO: add logging output here
-    std::cout << "Found property through path below:\n";
+bool ServerConfig::propertyLookUp(const std::vector<std::string>& propertyPath,
+                                  std::string& val) const {
+  std::unordered_map<std::vector<std::string>,
```
Comment (KyCodeHuynh) [inline server_config.cc]: And people rag on Java for verbosity... (not your fault; it's C++. xD)

File: server_config.cc [hunk 8]
```diff
@@ -41,83 +34,156 @@ std::string buildWordyParam(std::shared_ptr<NginxConfigStatement> statement,
...
-    // TODO: Possible logging here about empty config
-    return;
+    std::cout << "Failed to build ServerConfig map, recieved empty config\n";
+    return false;
   }
 
   // Loop through all config param block at the current depth
   for (size_t i = 0; i < config.statements_.size(); i++) {
-    size_t numTokens = config.statements_[i]->tokens_.size();
+    std::shared_ptr<NginxConfigStatement> curStatement = config.statements_[i];
```
Comment (kfrankc) [inline server_config.cc]: Like Ky-Cuong said, `snake_case` for variables, so `curStatement` should be changed to `cur_statement`?

File: server_config.cc [hunk 9]
```diff
@@ -41,83 +34,156 @@ std::string buildWordyParam(std::shared_ptr<NginxConfigStatement> statement,
...
+    size_t numTokens = curStatement->tokens_.size();
```
Comment (kfrankc) [inline server_config.cc]: Same with `numTokens`, should be `num_tokens`

File: server_config.cc [hunk 10]
```diff
@@ -41,83 +34,156 @@ std::string buildWordyParam(std::shared_ptr<NginxConfigStatement> statement,
...
+    }
   }
+
+  std::unique_ptr<NginxConfig> emptyConfig =
+    std::unique_ptr<NginxConfig>(new NginxConfig());
+  return emptyConfig;
 }
+
+// Print the contents of the mapping of 'path to config param' -> value
+void ServerConfig::printPropertiesMap() {
```
Comment (kfrankc) [inline server_config.cc]: If we continue with our style convention, this should be `PrintPropertiesMap()`

Comment (apfong) [inline server_config.cc]: I think we'd still have camel case here since its a function

File: server_config.cc (hunk 6 above)
Comment (apfong) [inline server_config.cc]: That is indeed what we mean

**Comment by apfong** [comment]
Comments addressed by the following commits: [96d36b7](https://github.com/UCLA-CS130/Mr.-Robot-et-al./pull/50/commits/96d36b7c2d6518eb8f33ad3dbd2595e443884537) and [abd1bce](https://github.com/UCLA-CS130/Mr.-Robot-et-al./pull/50/commits/abd1bcebc5e2de28b124832d306edd04c207b946)

**Comment by KyCodeHuynh** [comment]
PR #49 is in. This now needs to be rebased atop `lightning-refactor`. 😺

**Pull request merged**